File: app/src/main/java/com/robotnec/budget/core/persistence/dao/impl/TransactionDaoImpl.java
package com.robotnec.budget.core.persistence.dao.impl;

import com.robotnec.budget.app.util.Mapper;
import com.robotnec.budget.core.domain.operation.Transaction;
import com.robotnec.budget.core.persistence.BudgetDatabase;
import com.robotnec.budget.core.persistence.dao.AccountDao;
import com.robotnec.budget.core.persistence.dao.CategoryDao;
import com.robotnec.budget.core.persistence.dao.TransactionDao;
import com.robotnec.budget.core.persistence.schema.TransactionRecord;
import com.yahoo.squidb.data.SquidCursor;
import com.yahoo.squidb.sql.Property;
import com.yahoo.squidb.sql.Query;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * @author zak <zak@swingpulse.com>
 */
public class TransactionDaoImpl extends SquidbDaoTemplate<Transaction, TransactionRecord>
        implements TransactionDao {

    private final AccountDao accountDao;
    private final CategoryDao categoryDao;

    public TransactionDaoImpl(BudgetDatabase database,
                              AccountDao accountDao,
                              CategoryDao categoryDao) {
        super(database);
        this.accountDao = accountDao;
        this.categoryDao = categoryDao;
    }

    @Override
    public List<Transaction> getTransactionsForCategory(long categoryId) {
        Query query = Query.select().where(TransactionRecord.CATEGORY_ID.eq(categoryId));
        SquidCursor<TransactionRecord> cursor = database.query(TransactionRecord.class, query);
        try {
            if (cursor.moveToFirst()) {
                List<TransactionRecord> result = new ArrayList<>();
                do {
                    result.add(fromCursor(cursor));
                } while (cursor.moveToNext());
                return map(result);
            }
        } finally {
            cursor.close();
        }
        return Collections.emptyList();
    }

    @Override
    TransactionRecord fromCursor(SquidCursor<TransactionRecord> cursor) {
        TransactionRecord record = new TransactionRecord();
        record.readPropertiesFromCursor(cursor);
        return record;
    }

    @Override
    Class<TransactionRecord> getRecordClass() {
        return TransactionRecord.class;
    }

    @Override
    List<Transaction> map(List<TransactionRecord> tableModels) {
        return Mapper.fromTransactionRecords(tableModels,
                accountDao::findById,
                categoryDao::findById);
    }

    @Override
    TransactionRecord map(Transaction item) {
        return Mapper.toRecord(item);
    }

    @Override
    Transaction map(TransactionRecord record) {
        return Mapper.fromRecord(record,
                accountDao::findById,
                categoryDao::findById);
    }

    @Override
    Property.LongProperty getIdProperty() {
        return TransactionRecord.ID;
    }
}

File: app/src/main/java/com/robotnec/budget/core/persistence/dao/impl/TransactionDaoImpl.kt
package com.robotnec.budget.core.persistence.dao.impl

import com.robotnec.budget.app.util.Mapper
import com.robotnec.budget.core.domain.operation.Transaction
import com.robotnec.budget.core.persistence.BudgetDatabase
import com.robotnec.budget.core.persistence.dao.AccountDao
import com.robotnec.budget.core.persistence.dao.CategoryDao
import com.robotnec.budget.core.persistence.dao.TransactionDao
import com.robotnec.budget.core.persistence.schema.TransactionRecord
import com.yahoo.squidb.data.SquidCursor
import com.yahoo.squidb.sql.Property
import com.yahoo.squidb.sql.Query
import java.util.ArrayList

/**
 * @author zak zak@swingpulse.com>
 */
class TransactionDaoImpl(database: BudgetDatabase,
                         private val accountDao: AccountDao,
                         private val categoryDao: CategoryDao)
    : SquidbDaoTemplate<Transaction, TransactionRecord>(database), TransactionDao {

    override fun getTransactionsForCategory(categoryId: Long): List<Transaction> {
        val query = Query.select().where(TransactionRecord.CATEGORY_ID.eq(categoryId))
        val cursor = database.query(TransactionRecord::class.java, query)
        try {
            if (cursor.moveToFirst()) {
                val result = ArrayList<TransactionRecord>()
                do {
                    result.add(fromCursor(cursor))
                } while (cursor.moveToNext())
                return map(result)
            }
        } finally {
            cursor.close()
        }
        return emptyList()
    }

    override fun fromCursor(cursor: SquidCursor<TransactionRecord>): TransactionRecord {
        val record = TransactionRecord()
        record.readPropertiesFromCursor(cursor)
        return record
    }

    override val recordClass: Class<TransactionRecord>
        get() = TransactionRecord::class.java

    override fun map(tableModels: List<TransactionRecord>): List<Transaction> {
        return Mapper.fromTransactionRecords(tableModels,
                { accountDao.findById(it) },
                { categoryDao.findById(it) })
    }

    override fun map(item: Transaction): TransactionRecord {
        return Mapper.toRecord(item)
    }

    override fun map(record: TransactionRecord): Transaction {
        return Mapper.fromRecord(record,
                { accountDao.findById(it) },
                { categoryDao.findById(it) })
    }

    override val idProperty: Property.LongProperty
        get() = TransactionRecord.ID
}
